### PR TITLE
Feature/access log filter

### LINKED ISSUE #750: Kubernetes ingress: allow to reduce request logging
Currently, ingress skipper produces log entry for each request and for some of the services we want to be able to switch off extensive request logging.

### PR DESCRIPTION
**Pull request by vecmezoni**
Fixes #750

* Removed access log http handler and moved logic to proxy 
* Added `accessLog` filter to give chance to override global AccessLogDisabled setting

### CODE REVIEW

**Comment by vecmezoni** [comment]
Can be considered just as proof of concept because I'm not sure that `proxy` is a good place for it.

**Comment by aryszka** [comment]
i like the idea, it's aligned with other features like the preserve host.

please, use `make fmt` before pushing. It's makes the build currently failing.

**Comment by vecmezoni** [comment]
@aryszka done

File: filters/accesslog/log_test.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var ctx filtertest.Context
+	ctx.FStateBag = make(map[string]interface{})
+
+	f.Request(&ctx)
+	bag := ctx.StateBag()
+	if bag[AccessLogEnabledKey] != false {
```
Comment (szuecs) [inline filters/accesslog/log_test.go]: can you also test the negative?
This test would also be fine if this is the default.

File: filters/accesslog/log.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+type accessLog struct {
+	status bool
+}
+
+// NewAccessLog creates a filter spec for overriding the state of the access log. (By default global setting is used.)
+//
+// 	accessLog("false")
+//
+// EXPERIMENTAL: this filter is experimental, and the name and the arguments can change until marked as stable.
```
Comment (aryszka) [inline filters/accesslog/log.go]: I would go as far as to not even mark it as experimental. @vecmezoni @szuecs , do you foresee any reason to change the behavior/name/args in the future?

Comment (szuecs) [reply filters/accesslog/log.go]: I don't see it no.

Comment (vecmezoni) [reply filters/accesslog/log.go]: I would think of a more generic approach when you can even provide log format here. Maybe it can be done with the separate filter.

File: filters/accesslog/log_test.go (hunk 1 above)
Comment (vecmezoni) [reply filters/accesslog/log_test.go]: sure, I'll update the test and also add the docs as you mentioned.

**Comment by vecmezoni** [comment]
@szuecs @aryszka 
So, I've renamed `accessLog` to `accessLogDisabled` to follow naming convention for Skipper setting `access-log-disabled` and remove possible confusion.
Added more tests and documentation.

File: filters/accesslog/disable.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+//
+// 	accessLogDisabled("false")
+func NewAccessLogDisabled() filters.Spec {
+	return &accessLogDisabled{}
+}
+
+func (al *accessLogDisabled) Name() string { return AccessLogDisabledName }
+
+func (al *accessLogDisabled) CreateFilter(args []interface{}) (filters.Filter, error) {
+	if len(args) < 1 {
```
Comment (szuecs) [inline filters/accesslog/disable.go]: `!= 1` shoud be used here

File: filters/accesslog/disable.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+		return nil, filters.ErrInvalidFilterParameters
+	}
+}
+
+func (al *accessLogDisabled) Request(ctx filters.FilterContext) {
+	bag := ctx.StateBag()
+	bag[AccessLogDisabledKey] = al.disabled
+}
+
+func (al *accessLogDisabled) Response(_ filters.FilterContext) {}
```
Comment (szuecs) [inline filters/accesslog/disable.go]: I think this can be used: `func (*accessLogDisabled) Response(filters.FilterContext) {}`

File: filters/accesslog/disable_test.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	}
+}
+
+func TestUnknownValue(t *testing.T) {
+	const state = "unknownValue"
+
+	_, err := NewAccessLogDisabled().CreateFilter([]interface{}{state})
+
+	if err == nil || err != filters.ErrInvalidFilterParameters {
+		t.Error("should throw error on unknown value")
```
Comment (szuecs) [inline filters/accesslog/disable_test.go]: throw -> return

File: filters/accesslog/disable_test.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+	bag := ctx.StateBag()
+	if bag[AccessLogDisabledKey] != true {
+		t.Error("failed to set access log state")
+	}
+}
+
+func TestUnknownValue(t *testing.T) {
+	const state = "unknownValue"
+
+	_, err := NewAccessLogDisabled().CreateFilter([]interface{}{state})
```
Comment (szuecs) [inline filters/accesslog/disable_test.go]:
```
if _, err := NewAccessLogDisabled().CreateFilter([]interface{}{state}); err == nil || err != filters.ErrInvalidFilterParameters {
...
}
```

File: filters/accesslog/disable_test.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+	ctx.FStateBag = make(map[string]interface{})
+
+	f.Request(&ctx)
+	bag := ctx.StateBag()
+	if bag[AccessLogDisabledKey] != false {
+		t.Error("failed to set access log state")
+	}
+}
+
+func TestDisabled(t *testing.T) {
```
Comment (szuecs) [inline filters/accesslog/disable_test.go]: If you like you can rewrite all these tests to table driven tests.
https://opensource.zalando.com/skipper/development/#writing-tests shows how to do table driven tests, which will shorten the code.

File: readme.md [hunk 8]
```diff
@@ -100,10 +100,10 @@ Getting the code with the test dependencies (`-t` switch):
...
 Build and test all packages:
 
-  cd src/github.com/zalando/skipper
-  make deps
-  make install
-  make shortcheck
+    cd src/github.com/zalando/skipper
+    make deps
+    make install
+    make shortcheck
```
Comment (szuecs) [inline readme.md]: please drop these changes

Comment (vecmezoni) [reply readme.md]: But it makes it look better:

default

File: filters/accesslog/disable.go (hunk 3 above)
Comment (vecmezoni) [reply filters/accesslog/disable.go]: 👍

File: filters/accesslog/disable.go (hunk 4 above)
Comment (vecmezoni) [reply filters/accesslog/disable.go]: 👍

File: filters/accesslog/disable_test.go (hunk 5 above)
Comment (vecmezoni) [reply filters/accesslog/disable_test.go]: 👍

File: filters/accesslog/disable_test.go (hunk 6 above)
Comment (vecmezoni) [reply filters/accesslog/disable_test.go]: made table tests, not needed anymore

File: filters/accesslog/disable_test.go (hunk 7 above)
Comment (vecmezoni) [reply filters/accesslog/disable_test.go]: 👍

**Comment by vecmezoni** [comment]
@szuecs done

File: readme.md (hunk 8 above)
Comment (szuecs) [reply readme.md]: Oh, I did not know, that it's not correctly rendered right now. :)

**Comment by szuecs** [comment]
👍

**Comment by vecmezoni** [comment]
👍

**Comment by aryszka** [comment]
:+1:

**Comment by mo-gr** [comment]
I know I'm too late to the party. But boolean flags are an interface smell. And abusing strings as boolean flags doesn't make the smell nicer. Have you considered to just turn this into two filters `accessLogDisabled()` and `accessLogEnabled()`? Also to avoid the unnecessary double negation of `accessLogDisabled("false")`...

**Comment by aryszka** [comment]
retrospectively, i agree. Didn't think of this when discussed the approach. Neither in case of this one https://godoc.org/github.com/zalando/skipper/filters/builtin#PreserveHost , long time ago.

**Comment by vecmezoni** [comment]
Unfortunatelly I haven’t considered this option. We can fix it until this gets wide usage. 

I was trying to follow configuration parameter name to remove confusion.

**Comment by szuecs** [comment]
@vecmezoni deprecate it and recreate the filter and split it into 2.

**Comment by szuecs** [comment]
Later we can drop everything we deprecate without breaking

**Comment by aryszka** [comment]
we can use new names: disableAccessLog() and enableAccessLog() for the new split filters, so that it doesn't conflict with the deprecated one.

**Comment by tkrop** [comment]
To late for the party, too. How about substituting the *true|false* concept against a real access log configuration, that allows to enable, disable, and in limits add information, e.g. a specific header. Disabling than could be expressed by an *empty|nil* configuration.

**Comment by ruiaraujo** [comment]
@tkrop You should comment on the new PR
